### DISCUSSION TITLE
Why Sirius web doesn't use EMF's command stack?

### DISCUSSION THREAD

**Discussion by luxingxiao** [eclipse-sirius/sirius-components#3252] [Q&A] [answered]
Can REDO/UNDO functions be implemented through the EMF command stack?

**Comment by pcdavid** [comment]
Undo/redo is not in our priorities at the moment.
It's actually non-trivial in a client/server and potentially collaborative context where multiple users (or the same user in different tabs/windows) are modifying the same project/share the same "command stack". A "naive" implementation would have potentially very confusing effects.

**Reply by luxingxiao** [reply]
The implementation of the Undo/redo function is indeed very complicated, but in comparison, the problems caused by the user's misoperation and the inability to undo seem to be more serious. For example, when the user accidentally deletes or data is modified due to misoperation, the data will be lost because the system does not support the undo operation. This loss will be very serious.

**Reply by luxingxiao** [reply]
Perhaps the system can add a protection mechanism, where the user decides to back up the finalized model. When the user makes a mistake, the data can be retrieved through historical versions. Regardless, a modeling system without data protection is very worrisome to use.

**Reply by pcdavid** [reply, marked as answer]
It's not a complete solution, but the upcoming 2024.3.0 version adds [a confirmation dialog](https://github.com/eclipse-sirius/sirius-web/issues/3089) for all deletion operations to reduce the risks of data loss.
